Use WeakHashSet<ScrollableArea> instead of HashSet<ScrollableArea*>

nmoucht opened the pull request:
#### 90f1d32abbcef785e1ae4b21fa131346c87f2fb1
<pre>
Use WeakHashSet&lt;ScrollableArea&gt; instead of HashSet&lt;ScrollableArea*&gt;
<a href="https://bugs.webkit.org/show_bug.cgi?id=249428">https://bugs.webkit.org/show_bug.cgi?id=249428</a>
rdar://103418459

Reviewed by Ryosuke Niwa.

Use WeakHashSet&lt;ScrollableArea&gt; instead of HashSet&lt;ScrollableArea*&gt;.
Make ScrollableArea compatible with CheckedPtr.

* Source/WebCore/dom/Document.cpp:
(WebCore::serviceScrollAnimationForScrollableArea):
(WebCore::Document::runScrollSteps):
* Source/WebCore/page/FocusController.cpp:
(WebCore::FocusController::setIsVisibleAndActiveInternal):
* Source/WebCore/page/FrameView.cpp:
(WebCore::FrameView::invalidateScrollbarsForAllScrollableAreas):
(WebCore::FrameView::notifyScrollableAreasThatContentAreaWillPaint const):
(WebCore::FrameView::addScrollableAreaForAnimatedScroll):
(WebCore::FrameView::removeScrollableAreaForAnimatedScroll):
(WebCore::FrameView::addScrollableArea):
(WebCore::FrameView::removeScrollableArea):
(WebCore::FrameView::containsScrollableArea const):
* Source/WebCore/page/FrameView.h:
* Source/WebCore/page/Page.cpp:
(WebCore::Page::lockAllOverlayScrollbarsToHidden):
(WebCore::Page::stopKeyboardScrollAnimation):
* Source/WebCore/page/scrolling/ScrollingCoordinator.cpp:
(WebCore::ScrollingCoordinator::absoluteEventTrackingRegionsForFrame const):
* Source/WebCore/platform/ScrollableArea.cpp:
* Source/WebCore/platform/ScrollableArea.h:
* Source/WebCore/testing/Internals.cpp:
(WebCore::Internals::numberOfScrollableAreas):
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::pageContainsAnyHorizontalScrollbars):

Canonical link: <a href="https://commits.webkit.org/258255@main">https://commits.webkit.org/258255@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/dde4f0b2b0e0d3c2932d870ad0384488e24af0fa

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/101140 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/10288 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/34180 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/110425 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/170673 "Built successfully and passed tests") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/105127 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/11225 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/1158 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/93533 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/108255 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/106922 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/8560 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/91784 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/35089 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/90432 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/23184 "Passed tests") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/78084 "Found 1 new API test failure: /WebKitGTK/TestWebKitAccessibility:/webkit/WebKitAccessibility/accessible/event-listener (failure)") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/3936 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/24676 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/3981 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/84/builds/1115 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/10083 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/44186 "Passed tests") | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/5736 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/75/builds/2975 "Built successfully and passed tests") | | | | 
<!--EWS-Status-Bubble-End-->

rniwa commented on Source/WebCore/dom/Document.cpp:
```diff
@@ -4489,10 +4489,15 @@ void Document::runScrollSteps()
         MonotonicTime now = MonotonicTime::now();
         bool scrollAnimationsInProgress = serviceScrollAnimationForScrollableArea(frameView.get(), now);
         HashSet<ScrollableArea*> scrollableAreasToUpdate;
```
Can we use HashSet of Ref or RefPtr here?

rniwa replied:
Oh, I guess ScrollableArea doesn't support Ref/RefPtr.
Why don't we make it compatible with CheckedPtr (`CanMakeCheckedPtr`) and use that instead.

nmoucht replied:
There is a static check to keep ScrollableArea the same size. @smfr do we have any problems with increasing ScrollableArea's size to be compatible with checked pointer?

nmoucht replied:
This only adds 16 bytes so it is ok to do this.

rniwa commented on Source/WebCore/page/Page.cpp:
```diff
@@ -1507,12 +1507,12 @@ void Page::lockAllOverlayScrollbarsToHidden(bool lockOverlayScrollbars)
 
         for (auto& scrollableArea : *scrollableAreas)
-            scrollableArea->lockOverlayScrollbarStateToHidden(lockOverlayScrollbars);
+            scrollableArea.lockOverlayScrollbarStateToHidden(lockOverlayScrollbars);
```
Ditto

nmoucht replied:
Fixed.

rniwa commented on Source/WebCore/page/FrameView.cpp:
```diff
@@ -820,8 +820,8 @@ void FrameView::invalidateScrollbarsForAllScrollableAreas()
-    for (auto* scrollableArea : *m_scrollableAreas)
-        scrollableArea->invalidateScrollbars();
+    for (auto& scrollableArea : *m_scrollableAreas)
+        scrollableArea.invalidateScrollbars();
```
Ditto.

nmoucht replied:
Fixed.

rniwa commented on Source/WebCore/page/FrameView.cpp:
```diff
@@ -4446,7 +4446,7 @@ void FrameView::notifyScrollableAreasThatContentAreaWillPaint() const
         // ScrollView ScrollableAreas will be handled via the Frame tree traversal above.
         if (!is<ScrollView>(scrollableArea))
-            scrollableArea->contentAreaWillPaint();
+            scrollableArea.contentAreaWillPaint();
```
Ditto

nmoucht replied:
Fixed.

rniwa commented on Source/WebCore/page/FocusController.cpp:
```diff
@@ -963,13 +963,13 @@ void FocusController::setIsVisibleAndActiveInternal(bool contentIsVisible)
-            ASSERT(scrollableArea->scrollbarsCanBeActive() || m_page.shouldSuppressScrollbarAnimations());
-            contentAreaDidShowOrHide(scrollableArea, contentIsVisible);
+            ASSERT(scrollableArea.scrollbarsCanBeActive() || m_page.shouldSuppressScrollbarAnimations());
+            contentAreaDidShowOrHide(&scrollableArea, contentIsVisible);
```
Can we store it in Ref/RefPtr before calling the function?

nmoucht replied:
Now store this in CheckedPtr.

rniwa commented on Source/WebCore/page/Page.cpp:
```diff
@@ -2563,8 +2563,8 @@ void Page::stopKeyboardScrollAnimation()
-            scrollableArea->stopAsyncAnimatedScroll();
-            scrollableArea->stopKeyboardScrollAnimation();
+            scrollableArea.stopAsyncAnimatedScroll();
+            scrollableArea.stopKeyboardScrollAnimation();
```
Ditto.

nmoucht replied:
Fixed.

rniwa commented on Source/WebCore/page/scrolling/ScrollingCoordinator.cpp:
```diff
@@ -126,11 +126,11 @@ EventTrackingRegions ScrollingCoordinator::absoluteEventTrackingRegionsForFrame(
         for (auto& scrollableArea : *scrollableAreas) {
             // Composited scrollable areas can be scrolled off the main thread.
-            if (!scrollableArea->isVisibleToHitTesting() || scrollableArea->usesAsyncScrolling())
+            if (!scrollableArea.isVisibleToHitTesting() || scrollableArea.usesAsyncScrolling())
```
Ditto

nmoucht replied:
Fixed.

rniwa commented on Source/WebKit/WebProcess/WebPage/WebPage.cpp:
```diff
@@ -6099,15 +6099,15 @@ static bool pageContainsAnyHorizontalScrollbars(Frame* mainFrame)
 
         for (auto* scrollableArea : *scrollableAreas) {
-            if (!scrollableArea->scrollbarsCanBeActive())
+            if (!scrollableArea.scrollbarsCanBeActive())
```
Ditto

nmoucht replied:
Fixed.

nmoucht committed:
Use WeakHashSet<ScrollableArea> instead of HashSet<ScrollableArea*>
https://bugs.webkit.org/show_bug.cgi?id=249428
rdar://103418459

Reviewed by Ryosuke Niwa.

Use WeakHashSet<ScrollableArea> instead of HashSet<ScrollableArea*>.
Make ScrollableArea compatible with CheckedPtr.

* Source/WebCore/dom/Document.cpp:
(WebCore::serviceScrollAnimationForScrollableArea):
(WebCore::Document::runScrollSteps):
* Source/WebCore/page/FocusController.cpp:
(WebCore::FocusController::setIsVisibleAndActiveInternal):
* Source/WebCore/page/FrameView.cpp:
(WebCore::FrameView::invalidateScrollbarsForAllScrollableAreas):
(WebCore::FrameView::notifyScrollableAreasThatContentAreaWillPaint const):
(WebCore::FrameView::addScrollableAreaForAnimatedScroll):
(WebCore::FrameView::removeScrollableAreaForAnimatedScroll):
(WebCore::FrameView::addScrollableArea):
(WebCore::FrameView::removeScrollableArea):
(WebCore::FrameView::containsScrollableArea const):
* Source/WebCore/page/FrameView.h:
* Source/WebCore/page/Page.cpp:
(WebCore::Page::lockAllOverlayScrollbarsToHidden):
(WebCore::Page::stopKeyboardScrollAnimation):
* Source/WebCore/page/scrolling/ScrollingCoordinator.cpp:
(WebCore::ScrollingCoordinator::absoluteEventTrackingRegionsForFrame const):
* Source/WebCore/platform/ScrollableArea.cpp:
* Source/WebCore/platform/ScrollableArea.h:
* Source/WebCore/testing/Internals.cpp:
(WebCore::Internals::numberOfScrollableAreas):
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::pageContainsAnyHorizontalScrollbars):

Canonical link: https://commits.webkit.org/258255@main
